Scope GitHub token to a list of provided private and public repository (#1244)
tektoncd/pipelines-as-code: docs/content/docs/guide/repositorycrd.md

- [Github Documentation for webhook events](https://docs.github.com/webhooks-and-events/webhooks/webhook-events-and-payloads?actionType=auto_merge_disabled#pull_request)
- [Gitlab Documentation for webhook events](https://docs.gitlab.com/ee/user/project/integrations/webhook_events.html)
{{< /hint >}}

## Scope GitHub token to a list of private and public repositories within and outside namespaces

> The proposal helps user to scope GitHub token to a list of provided repositories which exist in same namespace(By providing configuration at Repository level)

Reviewer: it's not a proposal :)

> The proposal helps user to scope GitHub token to a list of provided repositories which exist in same namespace(By providing configuration at Repository level)

Reviewer: please rephrase the documentaion.... doc folks are not going to redo everything for us

Author: Yes I am working on this

Author: Updated doc

as well as different namespaces(Global Configuration).

> It is useful in a scenario where CI Repos Differ from CD Repos, and

Reviewer: see my other comments i did this morning on the enhancement proposal, please try to phrase the documentation not like a proposal but as documentation, feel free to get help from chatgpt for english or ask to @mramendi for help

- [Github Documentation for webhook events](https://docs.github.com/webhooks-and-events/webhooks/webhook-events-and-payloads?actionType=auto_merge_disabled#pull_request)
- [Gitlab Documentation for webhook events](https://docs.gitlab.com/ee/user/project/integrations/webhook_events.html)
{{< /hint >}}

## Scope Github token to a list of private and public repositories within and outside namespaces

By default, the Github token that Pipelines as Code generates is scoped only to the repository where the payload comes from.
However, in some cases, the developer team might want the token to allow control over additional repositories.
For example, there might be a CI repository where the `.tekton/pr.yaml` file and source payload might be located, however the build process defined in `pr.yaml` might fetch tasks from a separate private CD repository.

You can extend the scope of the Github token in two ways:

- _Repository level configuration_: extend the Github token to a list of repositories that exist in the same namespace as the original repository.

> - _Global configuration_: extend the Github token to a list of repositories in different namespaces.

Reviewer: I think you may want to add that it is a admin configuration setting and the other is user setting (and reverse the order maybe starting with global and go after with repo level)

Author: Updated

### Note

Ignore Scoping of Github token for webhook based usecases.

### Prerequisite

- In the `pipelines-as-code` configmap, set the `secret-github-app-token-scoped` key to `false`.
This setting enables the scoping of the Github token to private and public repositories listed under the Global and Repository level configuration.

#### Scoping the Github token using Global configuration

> You can use global configuration to set a list of repositories in any namespaces.

Reviewer: ```suggestion
You can use the global configuration to use as a list of repositories used from any Repository CR in any namespaces.
```

To set the global configuration, in the `pipelines-as-code` configmap, set the `secret-github-app-scope-extra-repos` key, as in the following example:

>   ```yaml
>   apiVersion: v1
>   data:
>     secret-github-app-scope-extra-repos: "owner2/project2, owner3/project3"
>   kind: ConfigMap
>   metadata:
>     name: pipelines-as-code
>     namespace: pipelines-as-code
>   ```

Reviewer: ```suggestion
  ```yaml
  apiVersion: v1
  kind: ConfigMap
  metadata:
    name: pipelines-as-code
    namespace: pipelines-as-code
  data:
    secret-github-app-scope-extra-repos: "owner2/project2, owner3/project3"
  ```
```

You can use the `Repository` custom resource to scope the generated Github token to a list of repositories.
The repositories can be public or private, but must reside in the same namespace as the repository with which the `Repository` resource is associated.

Set the `github_app_token_scope_repos` spec configuration within the `Repository` custom resource, as in the following example:

  ```yaml
  apiVersion: "pipelinesascode.tekton.dev/v1alpha1"
  kind: Repository
  metadata:
    name: test
    namespace: test-repo
  spec:
    url: "https://github.com/linda/project"
    settings:
      github_app_token_scope_repos:
      - "owner/project"
      - "owner1/project1"
  ```

In this example, the `Repository` custom resource is associated with the `linda/project` repository in the `test-repo` namespace.
The scope of the generated Github token is extended to the `owner/project` and `owner1/project1` repositories, as well as the `linda/project` repository. These repositories must exist under the `test-repo` namespace.

**Note:**

If any of the repositories does not exist in the namespace, the scoping of the Github token fails with an error message as in the following example:

> ```yaml

Reviewer: ```suggestion
```console
```
